### DISCUSSION TITLE
Feature request: Add support for Sensirion SEN5x

### DISCUSSION THREAD

**Discussion by erdinguma** [arendst/Tasmota#15463] [Ideas]
Describe the problem you have/What new integration you would like

I'd like you to add support for the newest line of PM sensors from Sensirion - the SEN5x series.

Please describe your use case for this integration and alternatives you've tried:

Sensirion has released their newest line of air quality sensors, the SEN5x series. These sensors combine the best features of many of their other sensors, including the SPS30, SHT30, and more.

These sensors are also cheaper than Sensirion's SPS30 PM sensor (~$34 for the SEN54 vs. $55 for the SPS30 on DigiKey). I suspect most users will ultimately gravitate to the the SEN5x series as it offers more bang for the buck.

Adding this sensor would be a great addition to Tasmota!

Additional context

Relevant Links:
SEN55 - (PM, RH/T, VOC, NOx) - https://sensirion.com/products/catalog/SEN55/
SEN54 - (PM, RH/T, VOC) - https://sensirion.com/products/catalog/SEN54/
SEN50 - (PM) - https://sensirion.com/products/catalog/SEN50/

**Comment by sfromis** [comment]
It would be welcomed if "someone" submits a PR to include support for the SEN5x series.

**Reply by tyeth** [reply]
I welcome this too. I may attempt it as it has high value to me, but my approach previously was to use the Sensirion official driver and wrap that. Is that acceptable for tasmota?
-Edit- Added issue #17645

**Reply by sfromis** [reply]
Wrapping existing drivers is done for many cases of device support. Beware if the license clashes with open source terms, and binary blobs are not preferred, to put it mildly. And do note that if such a library gets big, it may not be a good fit for wrapping.

You should just not open an "issue" (which is for bugs) to request feature additions.

**Reply by tyeth** [reply]
Thanks, it wasn't clear how to proceed, apart from whats in the contrib docs etc. I'm also pants at c/c++ but can stumble through, so I guess the ticket was to ask for a bit of guidance too

**Reply by Jason2866** [reply]
Closed the issue and moved to discussion and locked as duplicated.
Clearly speaking issues are only for issues. Not a problem at all, please remember :-)

**Reply by Jason2866** [reply]
Just ask here what guidance or better concrete questions you have.

**Reply by tyeth** [reply]
Okay so for another project I directly included the sensirion drivers (arduino compatible versions) as library requirements (dependencies) using library.proprties file, and then wrote an appropriate wrapper. So in this project how should one include libraries, is it a case of just cloning a library into lib, then sticking in a driver file like it mentions here: https://tasmota.github.io/docs/Sensor-API/#directoryfile-structure
Driver I used: https://github.com/Sensirion/arduino-i2c-sen5x

**Reply by Jason2866** [reply]
Yes, every lib is copied! in Tasmota repo. Since it is a i2c lib it has to be placed in lib folder i2c.
Using the sensirion lib is okay since it uses BSD 3 license.

**Reply by tyeth** [reply]
There's a slight difference in how the SCD40 and SGP40 commit changes to sensor values. Does the global version like sgp40 does assume only one sensor possibly connected, and is therefore unsafe if the user plans to have multiple temp/humidity related sensors attached? ```sen5x.measureRaw(TasmotaGlobal.temperature_celsius, TasmotaGlobal.humidity);``` [read sen5x as sgp40, replaced copypasta]
They both do the update function which seems to be how it is preferred:
<img width="272" alt="image" src="https://user-images.githubusercontent.com/6692083/211556403-198ac885-f8d7-40db-86f7-e229600d68e8.png"> 
<img width="529" alt="image" src="https://user-images.githubusercontent.com/6692083/211556562-c4f1f871-2083-4f17-972b-adcb3f241d84.png">
Think I'm about to run into https://github.com/arendst/Tasmota/issues/15834 and https://github.com/arendst/Tasmota/issues/12630 and possibly https://github.com/arendst/Tasmota/discussions/17277 (SGP30/40 issue) depending on what tech debt I inherit.
I assume safest is to not update the global options at all, but I want to be useful to most people.
Personally I will have an SCD41 connected too, and potentially other temp/humidity devices (I've got no SGP's, but the SEN5X is equivalent, and have SHT2/3/4/CX AHT20 DS18b20 SCD41s).
Do I update the global parameters? Do I go with the SCD40 driver as skeleton code instead of SGP40, as it offers the none-global option, although I've no desire to offer all the sensor commands functionality? Do I refactor the SGP40 to support the SEN5X in one driver file (mostly so I can reuse the 69 xns number which matches the i2c address 😆 )?

**Reply by Jason2866** [reply]
Maybe making global option configurable via  `Sensorxx` command https://tasmota.github.io/docs/Sensor-API/#important-things-to-consider Default setting not global. So it is up to the user.

**Reply by arendst** [reply]
Don't worry about the global temp/hum/pressure values. It has totally been rewritten recently and is now under the control of the user. Once all sensors are connected and a stable SENSOR JSON has been built the user just selects a global temperature as index to the JSON message. So looking at the message counting from the amount of TEMPERATURE labels in the message the user just selects one of them as being the global temperature to act upon.

So in short you can rely on a fixed global temperature (as long as you select one out of the available connected temperature sensors). Ditto for humidity and pressure. Look for commands like ``GlobalTemp`` and ``GlobalHum``

**Reply by tyeth** [reply]
Think I get you. So if I've added JSON and web output, then all should be well as the user can choose any of the json parameteres for global usage? I'm also reusing the temp/humidty helper for printing, which maybe is legacy and a bad idea? (line 229)
https://github.com/tyeth/Tasmota/commit/454ee95702495474b9a270c8a1c7b66cdea40baf#diff-5927bb987c5015b1aaaf4aa24bafb8b006d90fe5dd7a0d2288176ad7e1e9ee14R218-R233

**Reply by tyeth** [reply]
@erdinguma There's a pull request in, you can test it if you can build locally or on gitpod etc. The USE_SEN5X define needs to be set in your custom user overrides to enable the sensor.
https://github.com/arendst/Tasmota/pull/17736

**Reply by barbudor** [reply]
The pull request is incomplete
No driver inside

**Reply by tyeth** [reply]
Sorry about that 😴  should be there now

**Reply by DouglasRudd** [reply]
I am only getting readings for temp, humidity, abs humidity, and dewpoint, no pm or other sensor data from sen55

**Reply by sfromis** [reply]
Are you using latest dev binary? Are you seeing the data presented as "SEN5X"?

I can see that the current [driver](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino#L224-L272) will **always** include PM data when reporting.

**Reply by tyeth** [reply]
> Are you using latest dev binary? Are you seeing the data presented as "SEN5X"?
> 
> I can see that the current [driver](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino#L224-L272) will **always** include PM data when reporting.

Yeah that's the case for all models, so the 50 has PM data, but only the 54+55 have temp/humidity/VOC, and 55 also has NOX

**Reply by DouglasRudd** [reply]
yes I am using the latest dev Binary.
I also notice the units dont match either. attached is a screen shot of the web interface 

![sen55tas](https://user-images.githubusercontent.com/7246015/217665667-19891046-d830-4f4d-99b9-e0904277a61d.PNG)

and this is the output from the console.
SENSOR = {"Time":"2023-02-08T16:26:20","SEN5X":{"Temperature":22.6,"Humidity":24.6,"DewPoint":1.3,"aHumidity":4.9426},"TempUnit":"C"}

**Reply by tyeth** [reply]
Interesting :) Which model of the sensor, and which microcontroller / tasmota device? (Also which other sensors?)
If you compile with the debug sensors define turned on then it should show something mildly useful in the Console like this https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino#L109
I haven't really stuck in much error handling, and there is a weekly fan auto-clean which runs for ten minutes where readings will just fail silently in theory (untested).

**Reply by DouglasRudd** [reply]
I am still new with tasmota. so It could be a config issue.
sensor is the sen 55 no other sensors on this mcu
mcu is a esp8266ex.
00:00:00.002 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F4, Count 51
00:00:00.056 QPC: Count 1
00:00:00.070 ROT: Mode 1
00:00:00.071 SRC: Restart
00:00:00.073 Project tasmota - Lab _Sen55 Version 12.3.1.6(tasmota)-2_7_4_9(2023-02-08T22:18:25)
00:00:01.368 I2C: SEN5X found at 0x69
19:01:07.939 Running readMeasuredValues for SEN5X...
19:01:07.964 SEN5x readings:-
19:01:07.966 MassConcentrationPm1p0: 
19:01:07.968 MassConcentrationPm2p5: 
19:01:07.970 MassConcentrationPm4p0: 
19:01:07.971 MassConcentrationPm10p0: 
19:01:07.973 AmbientHumidity: 
19:01:07.975 AmbientTemperature: 
19:01:07.977 VocIndex: 
19:01:07.979 NoxIndex: n/a
19:01:07.981 AbsoluteHumidity:

**Reply by tyeth** [reply]
Aha, thanks for the extra info, fascinating! I'm also relatively new to it, and haven't tried an esp8266 yet. I wonder if I need to do something a little different, but in theory the Wire.begin should still work, which means the sensor connected on the first i2c bus. There is only support for Wire1 (second i2c bus) if ESP32 is defined. Only other thought is data types, which are floats and int16's as far as I'm aware which I naively assume is fine on esp8266. Possibly the backend sensirion library maybe doesn't support esp8266. I'll consider tasmoterising one of my smart-plugs to test the esp8266 over the weekend. Also your console isn't showing the extra debug info, I think you have to do a build with the extra [#define DEBUG_TASMOTA_SENSOR](https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L1194) which may then show a specific failure in the console.

**Reply by DouglasRudd** [reply]
I can try with an esp32, I checked the debug, and no errors.. I also flashed it with the sample Arduino code just the verify the sensor is working correctly

**Reply by tyeth** [reply]
You should see an additional message if the DEBUG_TASMOTA_SENSOR is defined, it will mention i2c bus number 0, either as a failure or success (Line 109 or 114 depending on error/success), may require setting logging to debug or more-debug in the console. 
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino#L109-L114
<img width="512" alt="image" src="https://user-images.githubusercontent.com/6692083/218136426-c96875ce-2460-460a-bb88-baa839d1ee69.png">

**Reply by DouglasRudd** [reply]
I do have the debug set and the output you can see in my previous post, the sensor is detected on the correct address.

00:00:00.073 Project tasmota - Lab _Sen55 Version 12.3.1.6(tasmota)-2_7_4_9(2023-02-08T22:18:25)
**00:00:01.368 I2C: SEN5X found at 0x69**
19:01:07.939 Running readMeasuredValues for SEN5X...

both option 3 and for (debug and debug more) to not show any errors.

**Comment by tyeth** [comment]
Sorry for the lack of explanation on my part, the line you highlight is
tasmota executing line 113, not debug sensor output, and tasmota doesn't
log the bus number to console (although it is supplied to the I2cSetActive
function), and the lines I mentioned (109 or 114) do mention a bus, so
there should be a second similar message ending with "bus 0" logged to
console.

**Reply by DouglasRudd** [reply]
I think we are looking in the wrong direction, I think it has something to do with the float conversion to string/char in the PTRS().
I have found that it will not output with a %f, I also found in the docs reference to %_f (https://tasmota.github.io/docs/Sensor-API), and looking at some of the other code, I found the use of dtosrtfd() to convert floats to char and then using %s for output. I am testing this theory now.
what compliler are you using? I am using gitpod.

**Reply by DouglasRudd** [reply]
FYI here is the output from debug RE the Bus info
00:00:00.073 Project tasmota - Sen55 Version 12.3.1.6(tasmota)-2_7_4_9(2023-02-12T01:36:50)
00:00:01.368 I2C: SEN5X found at 0x69
00:00:01.369 ************ Sensirion SEN5X found, i2c bus 0

**Comment by arendst** [comment]
Tasmota does not support %f as it's library is way too large. We have our own implementation by using %*_f where * may be an option or a fixed number of decimals. The variable also needs to be given as a pointer (superceded with &).

@tyeth some changes recommended:
- I suggest you take a look at any other driver searching for %*_f to fix.
- While going over the driver also revisit the use of hardcoded unit names. They should also be replaced by standard language specific unit names.
- Replace the fixed celsius unit by the temperature conversion unit making it available for fahrenheit too. Again search the code for examples of how to do this.

**Reply by tyeth** [reply]
You'll be relieved to hear it's over. Pull request #17961 
Testing and feedback gratefully appreciated @DouglasRudd @raptordemon , but locally tested using tasmota32s3cdc.

**Comment by raptordemon** [comment]
not sure if it has been mentioned but esphome supports this sensor, so there might be something that can be reviewed there: https://esphome.io/components/sensor/sen5x.html i have one on my desk

**Reply by tyeth** [reply]
Sorry it's taking me longer than desired to do the additional changes. I've [done the units and float changes as requested](https://github.com/tyeth/Tasmota/tree/fix-sen5x-floats) and I've got it displaying in console, but my c/c++ sucks, I'm getting LoadProhibited reboots after about 4reads, whereas before the %*_f changes it was working nicely for me. I'll muddle through, but my local environment is a mess and I'm not getting line numbers anymore, just corrupted stack message instead.

**Reply by tyeth** [reply]
Ah thank goodness. I deleted my users\tyeth\platformio folder (I'd pip installed previously) and have a clean enviroment. I still can't set breakpoints that work inside the driver, but it breaks on startup, and more importantly the exception has been caught, I can now progress easily having spotted my mistake.

**Comment by tyeth** [comment]
It's merged into development Thanks everyone for your contributions, many eyes appreciated 🦉

**Reply by bimac** [reply]
Thanks a lot for the effort! I just set two of these up on M5stamps - works beautifully.

**Reply by tyeth** [reply]
Very glad to hear it worked! Funnily enough its the new Ikea Air quality sensor dropping in April, SEN54, supporting matter with their hub via future update. Nice to know we've bagged a good'un sensor-wise

**Reply by bimac** [reply]
Uuh, thanks for the hint about the Ikea thingy - it looks quite promising! I'll definitely look into this. I set up two SEN55 for the office, but a zigbee version of the SEN54 would be great for home use.

**Reply by raptordemon** [reply]
Glad to hear you were able to make it work! Nicely done!

**Comment by bimac** [comment]
Should the "PM2.5" field perhaps be renamed to something like "PM2_5"? I think the dot could make parsing and handling a bit problematic in some cases.

**Reply by tyeth** [reply]
Fair point, no one _likes_ dots in their javascript field names...
Stricly speaking there is a less than symbol I wished to place beforehand, and I'm not a fan of PM for particulate matter, as it is easily confused for PPM when the unit is often ppb. The parsing and handling should be done in json-land not scraped from front page, and it's a legal fieldname so all is well if a little unpleasant for handling. I do half agree though, so I'm sorry for playing devils advocate.

**Reply by barbudor** [reply]
Other particule sensors also use the dot
It's indeed far from perfect for parsing but can be handled as `xxx['PM2.5']`

**Reply by Jason2866** [reply]
Yeah, remember had to do some trickery to get the dot parsed correctly in OpenHab

**Comment by noppingen** [comment]
Does anybody have a solution for this:

![Screenshot_20250412-214151](https://github.com/user-attachments/assets/03ecc11d-22ca-4088-b035-8b1387591e58)

Tried numerous build options, always the same result. PM values start at around 1000 after power on and then increase linear. The SEN54 is obviously detected ok. 

A buffer overrun maybe? Sensirions SPS30 needed an increased i2c buffer size earlier, is that somehow related maybe?

**Comment by noppingen** [comment]
Note to myself: Using an ESP32S2 with 240MHz and setting `Sleep 100` fixed the issue. I guess my WebQuery rules  left too little ressources for the loop.

**Reply by Jason2866** [reply]
Indeed it is stressing the one core S2 a lot "answering" WebQuerys. There is a reason why mqtt is the heart of Tasmota.
